File: app/src/main/java/fansirsqi/xposed/sesame/util/Detector.java
package fansirsqi.xposed.sesame.util;

import android.content.Context;
import android.content.pm.PackageManager;

import java.io.File;

import fansirsqi.xposed.sesame.BuildConfig;

public class Detector {


    public static String getLibPath(Context context) {
        String libSesamePath = null;
        try {
            libSesamePath = context.getPackageManager().
                    getApplicationInfo(BuildConfig.APPLICATION_ID, 0).
                    nativeLibraryDir + File.separator + System.mapLibraryName(
                    "checker");
        } catch (PackageManager.NameNotFoundException e) {
            ToastUtil.showToast(context, "请不要对应用宝隐藏TK模块");
            Log.record("请不要对应用宝隐藏TK模块");
            Log.error("getLibPath", e.getMessage());
        }
        return libSesamePath;
    }


    public static Boolean loadLibrary(String libraryName) {
        try {
            System.loadLibrary(libraryName);
            return true;
        } catch (UnsatisfiedLinkError e) {
            Log.error("loadLibrary", e.getMessage());
            return false;
        }
    }

    private static native void init(Context context);
    public static native void tips(Context context, String message);

    public static void initDetector(Context context) {
        try {
            init(context);
        } catch (Exception e) {
            Log.error("initDetector", e.getMessage());
        }
    }


    // 只保留一个 native 方法声明
    private static native boolean checkImpl(Context context);

    public static boolean isLspatchDetected(Context context) {
        try {
            return checkImpl(context);
        } catch (Exception e) {
            Log.error("isLspatchDetected", e.getMessage());
            return false;
        }
    }
}

File: app/src/main/java/fansirsqi/xposed/sesame/util/Detector.kt
package fansirsqi.xposed.sesame.util

import android.content.Context
import android.content.pm.PackageManager
import fansirsqi.xposed.sesame.BuildConfig
import java.io.File

object Detector {

    fun getLibPath(context: Context): String? {
    var libSesamePath: String? = null
    try {
        val appInfo = context.packageManager.getApplicationInfo(
            BuildConfig.APPLICATION_ID, 0
        )
        libSesamePath = appInfo.nativeLibraryDir + File.separator + System.mapLibraryName("checker")
        Log.runtime("load libSesame success")
    } catch (e: PackageManager.NameNotFoundException) {
        ToastUtil.showToast(context, "请不要对应用宝隐藏TK模块")
        Log.record("请不要对应用宝隐藏TK模块")
        Log.error("getLibPath", e.message)
    }
    return libSesamePath
}

    fun loadLibrary(libraryName: String): Boolean {
        try {
            System.loadLibrary(libraryName)
            return true
        } catch (e: UnsatisfiedLinkError) {
            Log.error("loadLibrary", e.message)
            return false
        }
    }

    private external fun init(context: Context)
    external fun tips(context: Context, message: String?)
    external fun isEmbeddedNative(context: Context): Boolean
    external fun dangerous(context: Context)

    /**
     * 检测是否通过LSPatch运行
     */
    private fun isRunningInLSPatch(context: Context): Boolean {
        try {
            // 检查应用元数据中是否有LSPatch标记
            val appInfo = context.packageManager.getApplicationInfo(
                context.packageName,
                PackageManager.GET_META_DATA
            )
            val hasLSPatch = appInfo.metaData?.containsKey("lspatch") == true
            return appInfo.metaData?.containsKey("lspatch") == true
        } catch (e: Exception) {
            Log.error("检查LSPatch运行环境时出错: ${e.message}")
            return false
        }
    }



    /**
     * 检测模块是否在合法环境中运行
     */
    fun isLegitimateEnvironment(context: Context): Boolean {
        val isRunningInLSPatch = isRunningInLSPatch(context)
        if (!isRunningInLSPatch) {
            return false
        }
        val isEmbedded = isEmbeddedNative(context)
        Log.runtime("isEmbedded: $isEmbedded")
        return isEmbedded
    }



    fun initDetector(context: Context) {
        try {
            init(context)
        } catch (e: Exception) {
            Log.error("initDetector", e.message)
        }
    }

    private fun getApkPath(context: Context, packageName: String): String? {
        try {
            val pm = context.packageManager
            val appInfo = pm.getApplicationInfo(packageName, 0)
            Log.runtime("appInfo.sourceDir: " + appInfo.sourceDir)
            return appInfo.sourceDir
        } catch (_: PackageManager.NameNotFoundException) {
            Log.runtime("Package not found: $packageName")
            return null
        }
    }

}